graylog2-web-interface/src/routing/App.jsx

```diff
@@ -1,54 +1,60 @@
 import PropTypes from 'prop-types';
 import React from 'react';
-import createReactClass from 'create-react-class';
-import Reflux from 'reflux';
+
 import Navigation from 'components/navigation/Navigation';
 import Spinner from 'components/common/Spinner';
 import Footer from 'components/layout/Footer';
+import connect from 'stores/connect';
+import StoreProvider from 'injection/StoreProvider';
+
+import AppErrorBoundary from './AppErrorBoundary';
 
 import 'stylesheets/jquery.dynatable.css';
 import 'stylesheets/typeahead.less';
 import 'c3/c3.css';
 import 'dc/dc.css';
 
-import StoreProvider from 'injection/StoreProvider';
-import AppErrorBoundary from './AppErrorBoundary';
-
 const CurrentUserStore = StoreProvider.getStore('CurrentUser');
 
-const App = createReactClass({
-  displayName: 'App',
-
-  propTypes: {
-    children: PropTypes.oneOfType([
-      PropTypes.arrayOf(PropTypes.element),
-      PropTypes.element,
-    ]).isRequired,
-    location: PropTypes.object.isRequired,
-  },
-
-  mixins: [Reflux.connect(CurrentUserStore)],
-
-  render() {
-    if (!this.state.currentUser) {
-      return <Spinner />;
-    }
-    return (
-      <div>
-        <Navigation requestPath={this.props.location.pathname}
-                    fullName={this.state.currentUser.full_name}
-                    loginName={this.state.currentUser.username}
-                    permissions={this.state.currentUser.permissions} />
-        <div id="scroll-to-hint" style={{ display: 'none' }} className="alpha80">
-          <i className="fa fa-arrow-up" />
-        </div>
-        <AppErrorBoundary>
-          {this.props.children}
-        </AppErrorBoundary>
-        <Footer />
+const App = ({ children, currentUser: { currentUser }, location }) => {
+  if (!currentUser) {
+    return <Spinner />;
+  }
+
+  return (
+    <div>
+      <Navigation requestPath={location.pathname}
+                  fullName={currentUser.full_name}
+                  loginName={currentUser.username}
+                  permissions={currentUser.permissions} />
+      <div id="scroll-to-hint" style={{ display: 'none' }} className="alpha80">
+        <i className="fa fa-arrow-up" />
       </div>
-    );
-  },
-});
-
-export default App;
+      <AppErrorBoundary>
+        {children}
+      </AppErrorBoundary>
+      <Footer />
+    </div>
+  );
+};
+
+App.propTypes = {
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.element),
+    PropTypes.element,
+  ]).isRequired,
+  currentUser: PropTypes.shape({
+    currentUser: PropTypes.shape({
+      full_name: PropTypes.string,
+      username: PropTypes.string,
+      permissions: PropTypes.array,
+    }),
+  }),
+  location: PropTypes.object.isRequired,
+};
+
+App.defaultProps = {
+  currentUser: undefined,
+};
+
+export default connect(App, { currentUser: CurrentUserStore });
```

Comment:
We can flatten the structure of the `currentUser` prop by using a props mapper here. This would reduce the clutter a bit with the nested `currentUser.currentUser` prop needing destructuring inside the component and makes it easier to replace the `connect` with a context later. You can do it like this:

```suggestion
export default connect(App, { currentUser: CurrentUserStore }, ({ currentUser: { currentUser } = {} }) => ({ currentUser }));
```

Reply:
I like the idea, one thought:

Would `currentUser: { currentUser } = {}` overwrite the `App.defaultProps`? In this case we would also need to modify the `if (!currentUser)` check.

Reply:
The default assignment is for the case when `connect` calls `getInitialState` on the `CurrentUserStore` when it has no state currently (which should never happen™, because this means that no user is logged in), so it is just a safe coding measure. In this case, the `currentUser` prop which is passed to the `App` component would be mapped to `undefined`, which is effectively the default prop. Does this answer the question?

Reply:
Yes, thank you

Reply:
@kyleknighted Just had a look together with Dennis and he mentioned, that in this context the above change is the only necessary one, because `({ currentUser: { currentUser } = {} }) => ({ currentUser }))` will not set the App prop currentUser to an empty object, it would still be undefined. I misread that one.

So the defaultProp for currentUser can still be undefined and the check can still be `if (!currentUser)` as well.
Sorry for the confusion.